Move to new error prone plugin

https://github.com/tbroyer/gradle-errorprone-javacplugin-plugin

Better compatibility with gradle moving forward, though not sure why the original wasn't just updated with these changes 🤔 

FYI @msridhar @kageiit

@tbroyer I've been giving this another spin in [this branch](https://github.com/uber/AutoDispose/tree/z/newEPPlugin), but continuing to see the errors from before. Any suggestions?

Example trace:

```
An exception has occurred in the compiler (1.8.0_131). Please file a bug against the Java compiler via the Java bug reporting page (http://bugreport.java.com) after checking the Bug Database (http://bugs.java.com) for duplicates. Include your program and the following diagnostic in your report. Thank you.
java.lang.RuntimeException: java.lang.BootstrapMethodError: java.lang.NoClassDefFoundError: com/sun/tools/javac/util/JavacMessages$ResourceBundleHelper
        at com.sun.tools.javac.main.Main.compile(Main.java:473)
        at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)
        at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)
        at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:50)
        at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:36)
        at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.delegateAndHandleErrors(NormalizingJavaCompiler.java:86)
        at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.execute(NormalizingJavaCompiler.java:47)
        at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.execute(NormalizingJavaCompiler.java:33)
        at org.gradle.api.internal.tasks.compile.CleaningJavaCompilerSupport.execute(CleaningJavaCompilerSupport.java:40)
        at org.gradle.api.internal.tasks.compile.CleaningJavaCompilerSupport.execute(CleaningJavaCompilerSupport.java:27)
        at org.gradle.api.tasks.compile.JavaCompile.performCompilation(JavaCompile.java:161)
        at org.gradle.api.tasks.compile.JavaCompile.compile(JavaCompile.java:146)
        at org.gradle.api.tasks.compile.JavaCompile.compile(JavaCompile.java:118)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        at org.gradle.api.internal.project.taskfactory.IncrementalTaskAction.doExecute(IncrementalTaskAction.java:50)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:39)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:26)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:131)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:317)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:309)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:185)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:97)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:120)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:99)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.OutputDirectoryCreatingTaskExecuter.execute(OutputDirectoryCreatingTaskExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:101)
        at org.gradle.api.internal.tasks.execution.FinalizeInputFilePropertiesTaskExecuter.execute(FinalizeInputFilePropertiesTaskExecuter.java:44)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:91)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.run(EventFiringTaskExecuter.java:51)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:317)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:309)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:185)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:97)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:46)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$ExecuteTaskAction.execute(DefaultTaskExecutionGraph.java:262)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$ExecuteTaskAction.execute(DefaultTaskExecutionGraph.java:246)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:136)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:130)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.execute(DefaultTaskPlanExecutor.java:201)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.executeWithTask(DefaultTaskPlanExecutor.java:192)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:130)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.BootstrapMethodError: java.lang.NoClassDefFoundError: com/sun/tools/javac/util/JavacMessages$ResourceBundleHelper
        at com.google.errorprone.BaseErrorProneJavaCompiler.setupMessageBundle(BaseErrorProneJavaCompiler.java:203)
        at com.google.errorprone.ErrorProneJavacPlugin.init(ErrorProneJavacPlugin.java:40)
        at com.sun.tools.javac.main.Main.compile(Main.java:470)
        ... 61 more
Caused by: java.lang.NoClassDefFoundError: com/sun/tools/javac/util/JavacMessages$ResourceBundleHelper
        ... 64 more
Caused by: java.lang.ClassNotFoundException: com.sun.tools.javac.util.JavacMessages$ResourceBundleHelper
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 64 more

```

--- a/android/autodispose-android-archcomponents-test/build.gradle
+++ b/android/autodispose-android-archcomponents-test/build.gradle
@@ -13,6 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+import net.ltgt.gradle.errorprone.CheckSeverity
 
 plugins {
   id 'com.android.library'
@@ -34,22 +35,24 @@ android {
 }
 
 dependencies {
-  annotationProcessor deps.build.nullAway
-  testAnnotationProcessor deps.build.nullAway
-  androidTestAnnotationProcessor deps.build.nullAway
-
   api project(':autodispose')
   api deps.androidx.annotations
   api deps.androidx.arch.lifecycle.common
   api deps.androidx.arch.lifecycle.runtime
 
+  errorproneJavac deps.build.errorProneJavac
+  errorprone deps.build.nullAway
   errorprone deps.build.errorProne
 }
 
-tasks.withType(JavaCompile) {
-  options.compilerArgs += ["-Xep:NullAway:ERROR",
-                           "-XepOpt:NullAway:AnnotatedPackages=com.uber",
-                           "-Xep:RestrictTo:OFF"]
+afterEvaluate {
+  tasks.withType(JavaCompile).configureEach {
+    options.errorprone {
+      option("NullAway:AnnotatedPackages", "com.uber")
+      check("NullAway", CheckSeverity.ERROR)
+      check("RestrictTo", CheckSeverity.OFF)
+    }
+  }
 }
 
 apply from: rootProject.file('gradle/gradle-mvn-push.gradle')

--- a/android/autodispose-android-archcomponents/build.gradle
+++ b/android/autodispose-android-archcomponents/build.gradle
@@ -13,6 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+import net.ltgt.gradle.errorprone.CheckSeverity
 
 plugins {
   id 'com.android.library'
@@ -43,9 +44,6 @@ android {
 }
 
 dependencies {
-  annotationProcessor deps.build.nullAway
-  testAnnotationProcessor deps.build.nullAway
-  androidTestAnnotationProcessor deps.build.nullAway
   annotationProcessor deps.androidx.arch.lifecycle.compiler
 
   api project(':autodispose')
@@ -56,6 +54,8 @@ dependencies {
   implementation project(':android:autodispose-android')
   implementation deps.rx.android
 
+  errorproneJavac deps.build.errorProneJavac
+  errorprone deps.build.nullAway
   errorprone deps.build.errorProne
 
   androidTestImplementation project(':android:autodispose-android-archcomponents-test')
@@ -69,10 +69,15 @@ dependencies {
   androidTestUtil deps.test.androidOrchestrator
 }
 
-tasks.withType(JavaCompile) {
-  options.compilerArgs += ["-Xep:NullAway:ERROR",
-                           "-XepOpt:NullAway:AnnotatedPackages=com.uber",
-                           "-Xep:RestrictTo:OFF"]
+afterEvaluate {
+  tasks.withType(JavaCompile).configureEach {
+    options.errorprone {
+      disableWarningsInGeneratedCode = true
+      option("NullAway:AnnotatedPackages", "com.uber")
+      check("NullAway", CheckSeverity.ERROR)
+      check("RestrictTo", CheckSeverity.OFF)
+    }
+  }
 }
 
 apply from: rootProject.file('gradle/gradle-mvn-push.gradle')

--- a/android/autodispose-android/build.gradle
+++ b/android/autodispose-android/build.gradle
@@ -13,6 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+import net.ltgt.gradle.errorprone.CheckSeverity
 
 plugins {
   id 'com.android.library'
@@ -38,15 +39,13 @@ android {
 }
 
 dependencies {
-  annotationProcessor deps.build.nullAway
-  testAnnotationProcessor deps.build.nullAway
-  androidTestAnnotationProcessor deps.build.nullAway
-
   api project(':autodispose')
   api deps.rx.java
   api deps.androidx.annotations
   implementation deps.rx.android
 
+  errorproneJavac deps.build.errorProneJavac
+  errorprone deps.build.nullAway
   errorprone deps.build.errorProne
 
   testImplementation project(':test-utils')
@@ -63,10 +62,14 @@ dependencies {
   androidTestUtil deps.test.androidOrchestrator
 }
 
-tasks.withType(JavaCompile) {
-  options.compilerArgs += ["-Xep:NullAway:ERROR",
-                           "-XepOpt:NullAway:AnnotatedPackages=com.uber",
-                           "-Xep:RestrictTo:OFF"]
+afterEvaluate {
+  tasks.withType(JavaCompile).configureEach {
+    options.errorprone {
+      option("NullAway:AnnotatedPackages", "com.uber")
+      check("NullAway", CheckSeverity.ERROR)
+      check("RestrictTo", CheckSeverity.OFF)
+    }
+  }
 }
 
 apply from: rootProject.file('gradle/gradle-mvn-push.gradle')

--- a/autodispose-rxlifecycle/build.gradle
+++ b/autodispose-rxlifecycle/build.gradle
@@ -13,33 +13,36 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+import net.ltgt.gradle.errorprone.CheckSeverity
 
 plugins {
-    id 'java-library'
-    id 'net.ltgt.errorprone'
+  id 'java-library'
+  id 'net.ltgt.errorprone'
 }
 
 sourceCompatibility = deps.build.javaVersion
 targetCompatibility = deps.build.javaVersion
 
 test {
-    testLogging.showStandardStreams = true
+  testLogging.showStandardStreams = true
 }
 
 dependencies {
-    annotationProcessor deps.build.nullAway
-    testAnnotationProcessor deps.build.nullAway
+  api project(':autodispose')
+  api deps.misc.rxlifecycle
 
-    api project(':autodispose')
-    api deps.misc.rxlifecycle
+  errorproneJavac deps.build.errorProneJavac
+  errorprone deps.build.nullAway
+  errorprone deps.build.errorProne
 
-    errorprone deps.build.errorProne
-
-    testCompile project(':test-utils')
+  testCompile project(':test-utils')
 }
 
-tasks.withType(JavaCompile) {
-    options.compilerArgs += ["-Xep:NullAway:ERROR", "-XepOpt:NullAway:AnnotatedPackages=com.uber"]
+tasks.withType(JavaCompile).configureEach {
+  options.errorprone {
+    option("NullAway:AnnotatedPackages", "com.uber")
+    check("NullAway", CheckSeverity.ERROR)
+  }
 }
 
 apply from: rootProject.file('gradle/gradle-mvn-push.gradle')

--- a/autodispose/build.gradle
+++ b/autodispose/build.gradle
@@ -13,6 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+import net.ltgt.gradle.errorprone.CheckSeverity
 
 plugins {
   id 'java-library'
@@ -27,22 +28,24 @@ test {
 }
 
 dependencies {
-  annotationProcessor deps.build.nullAway
-  testAnnotationProcessor deps.build.nullAway
-
   api deps.rx.java
 
+  errorproneJavac deps.build.errorProneJavac
+  errorprone deps.build.nullAway
   errorprone deps.build.errorProne
 
   testCompile project(':test-utils')
 }
 
-tasks.withType(JavaCompile) {
-  options.compilerArgs += ["-Xep:NullAway:ERROR", "-XepOpt:NullAway:AnnotatedPackages=com.uber"]
+tasks.withType(JavaCompile).configureEach {
+  options.errorprone {
+    option("NullAway:AnnotatedPackages", "com.uber")
+    check("NullAway", CheckSeverity.ERROR)
+  }
 }
 
 tasks.withType(Test) {
-    setMaxParallelForks(Runtime.runtime.availableProcessors().intdiv(2) ?: 1)
+  setMaxParallelForks(Runtime.runtime.availableProcessors().intdiv(2) ?: 1)
 }
 
 apply from: rootProject.file('gradle/gradle-mvn-push.gradle')

--- a/gradle/dependencies.gradle
+++ b/gradle/dependencies.gradle
@@ -19,7 +19,7 @@ def versions = [
     archLifecycle: '2.0.0',
     dokka: '0.9.17',
     errorProne: '2.3.2',
-    errorPronePlugin: '0.0.13',
+    errorPronePlugin: '0.6',
     kotlin: '1.3.10'
 ]
 

--- a/lifecycle/autodispose-lifecycle-jdk8/build.gradle
+++ b/lifecycle/autodispose-lifecycle-jdk8/build.gradle
@@ -13,6 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+import net.ltgt.gradle.errorprone.CheckSeverity
 
 plugins {
   id 'java-library'
@@ -27,18 +28,20 @@ test {
 }
 
 dependencies {
-  annotationProcessor deps.build.nullAway
-  testAnnotationProcessor deps.build.nullAway
-
   api project(":lifecycle:autodispose-lifecycle")
 
+  errorproneJavac deps.build.errorProneJavac
+  errorprone deps.build.nullAway
   errorprone deps.build.errorProne
 
   testCompile project(':test-utils')
 }
 
-tasks.withType(JavaCompile) {
-  options.compilerArgs += ["-Xep:NullAway:ERROR", "-XepOpt:NullAway:AnnotatedPackages=com.uber"]
+tasks.withType(JavaCompile).configureEach {
+  options.errorprone {
+    option("NullAway:AnnotatedPackages", "com.uber")
+    check("NullAway", CheckSeverity.ERROR)
+  }
 }
 
 apply from: rootProject.file('gradle/gradle-mvn-push.gradle')

--- a/lifecycle/autodispose-lifecycle/build.gradle
+++ b/lifecycle/autodispose-lifecycle/build.gradle
@@ -13,6 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+import net.ltgt.gradle.errorprone.CheckSeverity
 
 plugins {
   id 'java-library'
@@ -27,18 +28,20 @@ test {
 }
 
 dependencies {
-  annotationProcessor deps.build.nullAway
-  testAnnotationProcessor deps.build.nullAway
-
   api project(":autodispose")
 
+  errorproneJavac deps.build.errorProneJavac
+  errorprone deps.build.nullAway
   errorprone deps.build.errorProne
 
   testCompile project(':test-utils')
 }
 
-tasks.withType(JavaCompile) {
-  options.compilerArgs += ["-Xep:NullAway:ERROR", "-XepOpt:NullAway:AnnotatedPackages=com.uber"]
+tasks.withType(JavaCompile).configureEach {
+  options.errorprone {
+    option("NullAway:AnnotatedPackages", "com.uber")
+    check("NullAway", CheckSeverity.ERROR)
+  }
 }
 
 apply from: rootProject.file('gradle/gradle-mvn-push.gradle')

--- a/lifecycle/autodispose-lifecycle/src/test/java/com/uber/autodispose/lifecycle/LifecycleScopesTest.java
+++ b/lifecycle/autodispose-lifecycle/src/test/java/com/uber/autodispose/lifecycle/LifecycleScopesTest.java
@@ -125,7 +125,7 @@ public final class LifecycleScopesTest {
 
     @Override public boolean equals(Object o) {
       if (this == o) return true;
-      if (o == null || getClass() != o.getClass()) return false;
+      if (!(o instanceof IntHolder)) return false;
 
       IntHolder intHolder = (IntHolder) o;
 

--- a/static-analysis/autodispose-error-prone-checker/build.gradle
+++ b/static-analysis/autodispose-error-prone-checker/build.gradle
@@ -13,23 +13,17 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+import net.ltgt.gradle.errorprone.CheckSeverity
 
 plugins {
   id 'java-library'
   id 'net.ltgt.errorprone'
 }
 
-// we use this config to get the path of the JDK 9 javac jar, to
-// stick it in the bootclasspath when running tests
-configurations.maybeCreate("epJavac")
-
 sourceCompatibility = 1.8
 targetCompatibility = 1.8
 
 dependencies {
-  annotationProcessor deps.build.nullAway
-  testAnnotationProcessor deps.build.nullAway
-
   compileOnly deps.apt.autoService
   compileOnly deps.build.errorProneCheckApi
 
@@ -41,15 +35,20 @@ dependencies {
   testImplementation project(':autodispose')
   testImplementation project(':lifecycle:autodispose-lifecycle')
 
-  epJavac deps.build.errorProneJavac
+  errorproneJavac deps.build.errorProneJavac
+  errorprone deps.build.nullAway
+  errorprone deps.build.errorProne
 }
 
-tasks.withType(JavaCompile) {
-  options.compilerArgs += ["-Xep:NullAway:ERROR", "-XepOpt:NullAway:AnnotatedPackages=com.uber"]
+tasks.withType(JavaCompile).configureEach {
+  options.errorprone {
+    option("NullAway:AnnotatedPackages", "com.uber")
+    check("NullAway", CheckSeverity.ERROR)
+  }
 }
 
 test {
-  jvmArgs "-Xbootclasspath/p:${configurations.epJavac.asPath}"
+  jvmArgs "-Xbootclasspath/p:${configurations.errorproneJavac.asPath}"
 }
 
 apply from: rootProject.file("gradle/gradle-mvn-push.gradle")

--- a/test-utils/build.gradle
+++ b/test-utils/build.gradle
@@ -13,6 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+import net.ltgt.gradle.errorprone.CheckSeverity
 
 plugins {
   id 'java-library'
@@ -23,9 +24,8 @@ sourceCompatibility = deps.build.javaVersion
 targetCompatibility = deps.build.javaVersion
 
 dependencies {
-  annotationProcessor deps.build.nullAway
-  testAnnotationProcessor deps.build.nullAway
-
+  errorproneJavac deps.build.errorProneJavac
+  errorprone deps.build.nullAway
   errorprone deps.build.errorProne
 
   api deps.rx.java
@@ -33,6 +33,9 @@ dependencies {
   api deps.test.truth
 }
 
-tasks.withType(JavaCompile) {
-  options.compilerArgs += ["-Xep:NullAway:ERROR", "-XepOpt:NullAway:AnnotatedPackages=com.uber"]
+tasks.withType(JavaCompile).configureEach {
+  options.errorprone {
+    option("NullAway:AnnotatedPackages", "com.uber")
+    check("NullAway", CheckSeverity.ERROR)
+  }
 }

